### PR TITLE
Migrate NSGA2 sampler

### PR DESCRIPTION
**Pull request by not522**
## Motivation
Part of works for #1980. This PR migrates `NSGAIISampler` to the standard samplers module.
## Description of the changes
- add `optuna.samplers.NSGAIISampler` copied from `optuna.multi_objective.samplers.NSGAIIMultiObjectiveSampler`
- remove "MultiObjective" from sampler's name
- use ordinal `Study` and `FrozenTrial` in `NSGAIISampler`
- use `NSGAIISampler` as a default sampler for multi-objective study
- copy tests

### CODE REVIEW

**Comment by not522** [comment]
This PR changes the behavior of `examples/multi_objective/pytorch_simple.py`. I confirmed that it works.

**Comment by HideakiImamura** [comment]
@hvy Could you review this PR?

File: tests/samplers_tests/test_samplers.py [hunk 1]
```diff
@@ -35,6 +35,13 @@
         lambda: optuna.samplers.CmaEsSampler(n_startup_trials=0),
         lambda: optuna.integration.SkoptSampler(skopt_kwargs={"n_initial_points": 1}),
         lambda: optuna.integration.PyCmaSampler(n_startup_trials=0),
+        optuna.samplers.NSGAIISampler,
```
Comment (hvy) [inline tests/samplers_tests/test_samplers.py:38]: Great change 👍

**Review by hvy — COMMENTED** [review]
Thanks for the PR. Left some initial comments on the public API. I'll take another look at the internals and the tests tomorrow.

File: optuna/samplers/_nsga2.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        self._rng = np.random.RandomState(seed)
+
+    def reseed_rng(self) -> None:
+        self._random_sampler.reseed_rng()
+        self._rng = np.random.RandomState()
+
+    def infer_relative_search_space(
+        self,
+        study: Study,
+        trial: FrozenTrial,
```
Comment (HideakiImamura) [inline optuna/samplers/_nsga2.py:102]: A nit to reduce the number of lines.
```suggestion
        self, study: Study, trial: FrozenTrial
```

File: optuna/samplers/_nsga2.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+            parent_generation = generation
+            parent_population = population
+
+        return parent_generation, parent_population
+
+    def _select_elite_population(
+        self,
+        study: Study,
+        population: List[FrozenTrial],
```
Comment (HideakiImamura) [inline optuna/samplers/_nsga2.py:240]: Ditto.
```suggestion
        self, study: Study, population: List[FrozenTrial]
```

File: optuna/samplers/_nsga2.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                _crowding_distance_sort(population)
+                elite_population.extend(population[:n])
+                break
+
+        return elite_population
+
+    def _select_parent(
+        self,
+        study: Study,
+        population: List[FrozenTrial],
```
Comment (HideakiImamura) [inline optuna/samplers/_nsga2.py:258]: Ditto.
```suggestion
        self, study: Study, population: List[FrozenTrial]
```

File: optuna/samplers/_nsga2.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+        assert non_dominated_population
+        population_per_rank.append(non_dominated_population)
+
+    return population_per_rank
+
+
+def _crowding_distance_sort(
+    population: List[FrozenTrial],
+) -> None:
```
Comment (HideakiImamura) [inline optuna/samplers/_nsga2.py:314]: Ditto.
```suggestion
def _crowding_distance_sort(population: List[FrozenTrial]) -> None:
```

File: optuna/study.py [hunk 6]
```diff
@@ -733,8 +733,9 @@ def objective(trial):
 
         sampler:
             A sampler object that implements background algorithm for value suggestion.
-            If :obj:`None` is specified, :class:`~optuna.samplers.TPESampler` is used
-            as the default. See also :class:`~optuna.samplers`.
+            If :obj:`None` is specified, :class:`~optuna.samplers.TPESampler` is used for a
+            single-objective study and :class:`~optuna.samplers.NSGAIISampler` for a
+            multi-objective study. See also :class:`~optuna.samplers`.
```
Comment (HideakiImamura) [inline optuna/study.py:738]: To align other code bases, how about the following?
```suggestion
            If :obj:`None` is specified, :class:`~optuna.samplers.TPESampler` is used during
            single-objective optimization and :class:`~optuna.samplers.NSGAIISampler` during
            multi-objective optimization. See also :class:`~optuna.samplers`.
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. Basically, LGTM. I have some minor comments. PTAL.

File: optuna/samplers/_nsga2.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        crossover_prob:
+            Probability that a crossover (parameters swapping between parents) will occur
+            when creating a new individual.
+
+        swapping_prob:
+            Probability of swapping each parameter of the parents during crossover.
+
+        seed:
+            Seed for random number generator.
+
```
Comment (HideakiImamura) [inline optuna/samplers/_nsga2.py:60]: A nit.
```suggestion
```

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update! LGTM with a nit comment.

File: optuna/samplers/_nsga2.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+    def __init__(
+        self,
+        population_size: int = 50,
+        mutation_prob: Optional[float] = None,
+        crossover_prob: float = 0.9,
+        swapping_prob: float = 0.5,
+        seed: Optional[int] = None,
+    ) -> None:
+        # TODO(ohta): Reconsider the default value of each parameter.
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/samplers/_nsga2.py]: @not522, very sorry. I saw that you applied the experimental decorator but could you remove it again? On a second thought, it'll become the default sampler so it shouldn't raise a warning. Just double checked with @sile but it should be stable enough as well.

**Review by hvy — COMMENTED** [review]
LGTM besides some nits.

Also double checked with Kurobako to make sure results are consistent with [previous runs](https://github.com/optuna/optuna/pull/1163#issuecomment-620311938). 

zdt6-nsga-ii-32965fd1bfd4ea52694c539f74bd8d66a6ec10fefc54ab6bb501dbee0da36a7e
zdt5-nsga-ii-7a3564a416e74031a5b2a4ed93f10f191d7950e79bcf995255116395046d1189
zdt4-nsga-ii-6162ca3dea64314de9c00d600b0283c167c1f94c6398e9e8f9010da4d968ca85
zdt3-nsga-ii-6b7c781c95a74beefe75f003dfe36d48c2877a8d1b07b0c2aa1884af8e4d52f0
zdt2-nsga-ii-e8613148bc490a434243546011077c46cb2826d85b113fe0bada94baa18c6346
zdt1-nsga-ii-a7a1dae16638d36cba9c2589f9bfb644b5fe4cba90e83b77483b0ec7b6b0716c

**Comment by hvy** [comment]
@not522, could you just resolve the conflicts? Other than that, LGTM!

**Review by hvy — APPROVED** [review]
LGTM!

**Comment by hvy** [comment]
@not522, or were you finished with the changes?

**Comment by not522** [comment]
Yes, thank you for your reviews!

**Pull request merged**